### PR TITLE
Feature/102-Fix-flake8

### LINKED ISSUE #102: Fix flake8
Allow for some `# noqa`, specifically for complex functions/methods before we refactor those.

### PR DESCRIPTION
**Pull request by NoB0**
- Update files to solve flake8 issues
- Update some docstrings in file changed
- Reduce complexity of some methods
- Add #noqa to methods to be refactored

Methods to be refactored:
- `NLG.generate_output`
- `NLG._clarify_CIN`
- `NLG._user_options_remove_preference`
- `NLG._user_options_remove_preference_CIN`
- `UserIntentsChecker._filter_dact`
- `DataLoader.load_database`
- `NLU.generate_dact`
- `RBAnnotator._year_annotator`
- `DialoguePolicy.next_action`
- `DialogueStateTracker.update_state_user`

Fixes #102

### CODE REVIEW

File: docs/source/conf.py [hunk 1]
```diff
@@ -37,10 +37,10 @@
...
     "sphinx.ext.intersphinx",
     "sphinx.ext.napoleon",
-    #'sphinx.ext.coverage',  # in ll
-    #'sphinx.ext.ifconfig',
-    #'sphinx.ext.viewcode',
-    #'sphinx.ext.todo',
+    # 'sphinx.ext.coverage',  # in ll
+    # 'sphinx.ext.ifconfig',
+    # 'sphinx.ext.viewcode',
+    # 'sphinx.ext.todo',
```
Comment (IKostric) [inline docs/source/conf.py:43]: Nit: Remove commented (unused) lines

File: moviebot/agent/agent.py [hunk 2]
```diff
@@ -52,18 +50,18 @@ def _get_db(db_path):
...
-        :type self.bot_recorder: RecorderBot
-        :type self.dialogue_manager: DialogueManager
+    def __init__(self, config=None):
+        """Initializes the internal structure of the agent and other
+        components.
 
         Args:
-            config: (Default value = None)
-
+            config: Configuration.Defaults to None.
```
Comment (IKostric) [inline moviebot/agent/agent.py:64]: Nit: Add space between the sentences.

File: moviebot/agent/agent.py [hunk 3]
```diff
@@ -87,8 +85,7 @@ def initialize(self, user_id=None):
         configuration.
 
         Args:
-            user_id:  (Default value = None)
-
+            user_id:  User id.Defaults to None.
```
Comment (IKostric) [inline moviebot/agent/agent.py:88]: Nit: Add space between the sentences.

File: moviebot/agent/agent.py [hunk 4]
```diff
@@ -171,27 +165,21 @@ def start_dialogue(self, user_fname=None, restart=False):
...
         else:
-            record_data = (
-                self.dialogue_manager.dialogue_state_tracker.dialogue_state._dict()
-            )
+            record_data = self.dialogue_manager.get_state()._dict()
             record_data.update({"Agent_Output": agent_response})
             record_data.update(
                 {
-                    "Context": self.dialogue_manager.dialogue_state_tracker.dialogue_context.movies_recommended
+                    "Context": self.dialogue_manager.get_context().movies_recommended  # noqa
```
Comment (IKostric) [inline moviebot/agent/agent.py:182]: Nit: Add specific code for flake8.

File: moviebot/agent/agent.py [hunk 5]
```diff
@@ -229,28 +212,22 @@ def continue_dialogue(self, user_utterance, user_options, user_fname=None):
...
             record_data = {"User_Input": user_utterance.get_text()}
-            record_data.update(
-                self.dialogue_manager.dialogue_state_tracker.dialogue_state._dict()
-            )
+            record_data.update(self.dialogue_manager.get_state()._dict())
             record_data.update({"Agent_Output": agent_response})
             record_data.update(
                 {
-                    "Context": self.dialogue_manager.dialogue_state_tracker.dialogue_context.movies_recommended
+                    "Context": self.dialogue_manager.get_context().movies_recommended  # noqa
```
Comment (IKostric) [inline moviebot/agent/agent.py:230]: Nit: Add specific code for flake8.

File: moviebot/controller/controller_telegram.py [hunk 6]
```diff
@@ -448,12 +438,11 @@ def _recheck_user_options(self, options):
...
 
-    def _instruction(self, help=False):
-        """Instructions for new user
+    def _instruction(self, help: bool = False):
+        """Instructions for new user.
 
         Args:
-            help:  (Default value = False)
-
+            help: Whether to provide instructions or not.Defaults to False.
```
Comment (IKostric) [inline moviebot/controller/controller_telegram.py:445]: Nit: Add space between sentences.

File: moviebot/dialogue_manager/dialogue_policy.py [hunk 7]
```diff
@@ -19,32 +18,30 @@ class DialoguePolicy:
...
+            isBot: If the conversation is via bot or not.
         """
         self.ontology = ontology
         self.isBot = isBot
         self.new_user = new_user
 
-    def next_action(self, dialogue_state, dialogue_context=None, restart=False):
-        """Decides the next action to be taken by the agent based on the current
-        state and context.
+    def next_action(  # noqa
```
Comment (IKostric) [inline moviebot/dialogue_manager/dialogue_policy.py:32]: Nit: Add specific flake8 code (`# noqa: C901`)

File: moviebot/dialogue_manager/dialogue_state_tracker.py [hunk 8]
```diff
@@ -74,12 +72,11 @@ def _add_year_CIN(self, param):
                 value = value.replace("=", "").strip()
             self.dialogue_state.frame_CIN[param.slot] = value
 
-    def update_state_user(self, user_dacts):
-        """Updates the current dialogue state and context based on user DActs
+    def update_state_user(self, user_dacts):  # noqa
```
Comment (IKostric) [inline moviebot/dialogue_manager/dialogue_state_tracker.py:75]: Nit: Add specific flake8 code (`# noqa: C901`)

File: moviebot/nlg/nlg.py [hunk 9]
```diff
@@ -161,19 +156,18 @@ def __init__(self, args=None):
             Slots.YEAR.value: ["I couldn't find any timeline specification."],
         }
 
-    def generate_output(
-        self, agent_dacts, dialogue_state=None, user_fname=None
-    ):
+    def generate_output(  # noqa
```
Comment (IKostric) [inline moviebot/nlg/nlg.py:159]: Nit: Add specific flake8 code (`# noqa: C901`)

File: moviebot/nlg/nlg.py [hunk 10]
```diff
@@ -196,7 +190,7 @@ def generate_output(
                                 utterance.append(
                                     not_found_response.replace(
                                         "__replace__",
-                                        dialogue_state.user_utterance.get_text(),
+                                        dialogue_state.user_utterance.get_text(),  # noqa
```
Comment (IKostric) [inline moviebot/nlg/nlg.py:193]: Nit: Add specific flake8 code

File: moviebot/nlg/nlg.py [hunk 11]
```diff
@@ -341,12 +336,12 @@ def generate_output(
                     [
                         (
                             "Sorry, I don't have any "
-                            f'{"other " if dialogue_state.items_in_context else ""}'
+                            f'{"other " if dialogue_state.items_in_context else ""}'  # noqa
```
Comment (IKostric) [inline moviebot/nlg/nlg.py:339]: Nit: Add specific flake8 code

File: moviebot/nlg/nlg.py [hunk 12]
```diff
@@ -341,12 +336,12 @@ def generate_output(
...
                             f"{self._clarify_CIN(CIN, agent_dact)}."
                         ),
                         (
                             "Sorry, I couldn't find any "
-                            f'{"other " if dialogue_state.items_in_context else ""}'
+                            f'{"other " if dialogue_state.items_in_context else ""}'  # noqa
```
Comment (IKostric) [inline moviebot/nlg/nlg.py:344]: Nit: Add specific flake8 code

File: moviebot/nlg/nlg.py [hunk 13]
```diff
@@ -464,16 +459,15 @@ def _summarize_title_year(self, value):
         else:
             return f'year {"not " if negate else " "}' + value
 
-    def _clarify_CIN(self, CIN, agent_dact):
-        """Clarify the user CIN in the utterance
+    def _clarify_CIN(self, CIN, agent_dact) -> str:  # noqa
```
Comment (IKostric) [inline moviebot/nlg/nlg.py:462]: Nit: Add specific flake8 code (`# noqa: C901`)

File: moviebot/nlg/nlg.py [hunk 14]
```diff
@@ -631,15 +623,15 @@ def _user_options_recommend(self):
         options.update({"/restart": ["/restart"]})
         return options
 
-    def _user_options_remove_preference(self, dual_params):
-        """Generate options for user to select in case of two parameters have same value
+    def _user_options_remove_preference(self, dual_params):  # noqa
```
Comment (IKostric) [inline moviebot/nlg/nlg.py:626]: Nit: Add specific flake8 code (`# noqa: C901`)

File: moviebot/nlg/nlg.py [hunk 15]
```diff
@@ -804,15 +795,14 @@ def _user_options_inquire(self, dialogue_state):
         )
         return options
 
-    def _user_options_remove_preference_CIN(self, CIN):
-        """Generate options for user to select a parameter to remove
+    def _user_options_remove_preference_CIN(self, CIN):  # noqa
```
Comment (IKostric) [inline moviebot/nlg/nlg.py:798]: Nit: Add specific flake8 code (`# noqa: C901`)

File: moviebot/nlu/annotation/item_constraint.py [hunk 16]
```diff
@@ -1,8 +1,10 @@
 """The Dialogue Act Item defines the parameters for a Dialogue Act.
-It comprises ot the parameters mentioned in the conversation"""
 
+It comprises ot the parameters mentioned in the conversation
```
Comment (IKostric) [inline moviebot/nlu/annotation/item_constraint.py:3]: Nit: Missing dot.

File: moviebot/nlu/annotation/rule_based_annotator.py [hunk 17]
```diff
@@ -329,7 +326,7 @@ def _person_name_annotator(self, user_utterance, slots=None):
             if len(params) > 0:
                 return params
 
-    def _year_annotator(self, slot, user_utterance):
+    def _year_annotator(self, slot, user_utterance):  # noqa
```
Comment (IKostric) [inline moviebot/nlu/annotation/rule_based_annotator.py:329]: Nit: Add specific flake8 code

File: moviebot/nlu/data_loader.py [hunk 18]
```diff
@@ -59,13 +62,13 @@ def _convert_values_to_set(self, slot_values):
             else:
                 slot_values[slot] = set(values)
 
-    def load_database(self):
-        """Loads the database to fill dialogue slots with a list of possible slot_values.
-        This can be used further to understand what user intends to ask
+    def load_database(self):  # noqa
```
Comment (IKostric) [inline moviebot/nlu/data_loader.py:65]: Nit: Add specific flake8 code

File: moviebot/nlu/nlu.py [hunk 19]
```diff
@@ -1,62 +1,64 @@
...
-            config: Paths to ontology and database and tag words for slots in NLU
-
+            config: Paths to ontology, database and tag words for slots in NLU.
         """
         self.ontology = config["ontology"]
         self.database = config["database"]
         self.intents_checker = UserIntentsChecker(config)
 
-    def generate_dact(
+    def generate_dact(  # noqa
```
Comment (IKostric) [inline moviebot/nlu/nlu.py:42]: Nit: Add specific flake8 code (`# noqa: C901`)

File: moviebot/nlu/nlu.py [hunk 20]
```diff
@@ -172,7 +174,7 @@ def get_selected_option(self, user_utterance, options, item_in_focus):
                 isinstance(value, list) and value[0] == raw_utterance
             ) or value == raw_utterance:
                 if dact.intent == UserIntents.CONTINUE_RECOMMENDATION:
-                    dact.params = self.intents_checker.generate_params_continue_recommendation(
+                    dact.params = self.intents_checker.generate_params_continue_recommendation(  # noqa
```
Comment (IKostric) [inline moviebot/nlu/nlu.py:177]: Nit: Add specific flake8 code

File: moviebot/nlu/user_intents_checker.py [hunk 21]
```diff
@@ -502,16 +490,15 @@ def _remove_param(self, param, dact):
                 dact.params.remove(p)
                 return
 
-    def _filter_dact(self, dact, raw_utterance):
-        """This algorithm filters the DActs parameters and remove if one is
-        a sub-string of another. More filters are applied to remove the params
-        or change slots in a specified sequence if these qualify specific
-        conditions
+    def _filter_dact(self, dact, raw_utterance):  # noqa
```
Comment (IKostric) [inline moviebot/nlu/user_intents_checker.py:493]: Nit: Add specific flake8 code (`# noqa: C901`)

File: moviebot/database/database.py [hunk 22]
```diff
@@ -13,76 +16,88 @@
...
-        self.sql_connection = sqlite3.connect(
-            self.db_file_path
-        )  # SQL connection required to
-        # access the database
-        self.db_table_name = self._get_table_name()  # name of the table
+        # self.sql_connection = sqlite3.connect(
+        #     self.db_file_path
+        # )  # SQL connection required to
+        # # access the database
+        # self.db_table_name = self._get_table_name()  # name of the table
```
Comment (IKostric) [inline moviebot/database/database.py:36]: Remove commented lines.

File: moviebot/database/database.py [hunk 23]
```diff
@@ -13,76 +16,88 @@
...
-            the results of the SQL query
+            SQL condition if possible.
+        """
+        args = []
+        if dialogue_state.agent_should_offer_similar:
+            similar_movies = list(dialogue_state.similar_movies.values())[0]
+            if len(similar_movies) > 0:
+                for title in similar_movies:
+                    args.append(f'{Slots.TITLE.value} = "{title}"')
+                return " OR ".join(args) if len(args) > 0 else None
```
Comment (IKostric) [inline moviebot/database/database.py:64]: I am unsure about the implications here, but if `args` is empty `database_lookup` should return an empty list.

File: moviebot/controller/controller_messenger.py [hunk 24]
```diff
@@ -56,16 +57,16 @@ def __init__(self):
         self.start = {"get_started": {"payload": "/start"}}
 
     def get_started(self):
-        """Displays 'Get Started' button at messenger weclome screen"""
+        """Displays 'Get Started' button at messenger welcome screen."""
         return requests.post(
-            "https://graph.facebook.com/v10.0/me/messenger_profile?access_token="
+            "https://graph.facebook.com/v10.0/me/messenger_profile?access_token="  # noqa
```
Comment (IKostric) [inline moviebot/controller/controller_messenger.py:62]: Nit: Add specific flake8 code.

**Review by IKostric — CHANGES_REQUESTED** [review]
Two main points:
 * There is an issue with refactored methods in `database.py`
 * Add specific error codes we wish to ignore when using noqa

File: docs/source/conf.py (hunk 1 above)
Comment (NoB0) [reply docs/source/conf.py]: Done.

File: moviebot/agent/agent.py (hunk 2 above)
Comment (NoB0) [reply moviebot/agent/agent.py]: Done.

File: moviebot/agent/agent.py (hunk 3 above)
Comment (NoB0) [reply moviebot/agent/agent.py]: Done.

File: moviebot/agent/agent.py (hunk 5 above)
Comment (NoB0) [reply moviebot/agent/agent.py]: Done.

File: moviebot/agent/agent.py (hunk 4 above)
Comment (NoB0) [reply moviebot/agent/agent.py]: Done.

File: moviebot/dialogue_manager/dialogue_policy.py (hunk 7 above)
Comment (NoB0) [reply moviebot/dialogue_manager/dialogue_policy.py]: Done.

File: moviebot/controller/controller_telegram.py (hunk 6 above)
Comment (NoB0) [reply moviebot/controller/controller_telegram.py]: Done.

File: moviebot/nlg/nlg.py (hunk 9 above)
Comment (NoB0) [reply moviebot/nlg/nlg.py]: Done.

File: moviebot/nlg/nlg.py (hunk 10 above)
Comment (NoB0) [reply moviebot/nlg/nlg.py]: Done.

File: moviebot/nlg/nlg.py (hunk 12 above)
Comment (NoB0) [reply moviebot/nlg/nlg.py]: Done.

File: moviebot/nlg/nlg.py (hunk 13 above)
Comment (NoB0) [reply moviebot/nlg/nlg.py]: Done.

File: moviebot/nlg/nlg.py (hunk 14 above)
Comment (NoB0) [reply moviebot/nlg/nlg.py]: Done.

File: moviebot/nlg/nlg.py (hunk 11 above)
Comment (NoB0) [reply moviebot/nlg/nlg.py]: Done.

File: moviebot/nlg/nlg.py (hunk 15 above)
Comment (NoB0) [reply moviebot/nlg/nlg.py]: Done.

File: moviebot/nlu/annotation/item_constraint.py (hunk 16 above)
Comment (NoB0) [reply moviebot/nlu/annotation/item_constraint.py]: Done.

File: moviebot/nlu/annotation/rule_based_annotator.py (hunk 17 above)
Comment (NoB0) [reply moviebot/nlu/annotation/rule_based_annotator.py]: Done.

File: moviebot/nlu/data_loader.py (hunk 18 above)
Comment (NoB0) [reply moviebot/nlu/data_loader.py]: Done.

File: moviebot/nlu/nlu.py (hunk 19 above)
Comment (NoB0) [reply moviebot/nlu/nlu.py]: Done.

File: moviebot/nlu/nlu.py (hunk 20 above)
Comment (NoB0) [reply moviebot/nlu/nlu.py]: Done.

File: moviebot/database/database.py (hunk 22 above)
Comment (NoB0) [reply moviebot/database/database.py]: Done.

File: moviebot/nlu/user_intents_checker.py (hunk 21 above)
Comment (NoB0) [reply moviebot/nlu/user_intents_checker.py]: Done.

File: moviebot/database/database.py (hunk 23 above)
Comment (NoB0) [reply moviebot/database/database.py]: Good catch! I changed a bit the method `get_sql_condition` and added a if in the `database_lookup` to return an empty list if no similar movies were found.

**Review by IKostric — APPROVED** [review]
LGTM

**Comment by IKostric** [comment]
Can you make an issue to refactor the complex methods and copy the list of methods there?

**Pull request merged**